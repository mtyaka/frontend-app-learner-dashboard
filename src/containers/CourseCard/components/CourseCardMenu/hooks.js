import React from 'react';
import { StrictDict } from 'utils';
import * as module from './hooks';

export const state = StrictDict({
  isUnenrollConfirmVisible: (val) => React.useState(val), // eslint-disable-line
  isEmailSettingsVisible: (val) => React.useState(val), // eslint-disable-line
});

export const useUnenrollData = () => {
  const [isVisible, setIsVisible] = module.state.isUnenrollConfirmVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};

export const useEmailSettings = () => {
  const [isVisible, setIsVisible] = module.state.isEmailSettingsVisible(false);
  return {
    show: () => setIsVisible(true),
    hide: () => setIsVisible(false),
    isVisible,
  };
};
