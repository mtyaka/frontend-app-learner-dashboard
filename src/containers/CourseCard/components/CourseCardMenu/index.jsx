import React from 'react';
import PropTypes from 'prop-types';
import * as ReactShare from 'react-share';

import { useIntl } from '@edx/frontend-platform/i18n';
import { Dropdown, Icon, IconButton } from '@edx/paragon';
import { MoreVert } from '@edx/paragon/icons';

import { hooks as appHooks } from 'data/redux';
import EmailSettingsModal from 'containers/EmailSettingsModal';
import UnenrollConfirmModal from 'containers/UnenrollConfirmModal';
import { useEmailSettings, useUnenrollData } from './hooks';

import messages from './messages';

export const CourseCardMenu = ({ cardId }) => {
  const emailSettingsModal = useEmailSettings();
  const unenrollModal = useUnenrollData();
  const { courseName } = appHooks.useCardCourseData(cardId);
  const {
    // facebook,
    twitter,
  } = appHooks.useCardSocialSettingsData(cardId);
  const { isMasquerading } = appHooks.useMasqueradeData();
  const { formatMessage } = useIntl();

  return (
    <>
      <Dropdown>
        <Dropdown.Toggle
          id="dropdown-toggle-with-iconbutton"
          as={IconButton}
          src={MoreVert}
          iconAs={Icon}
          variant="primary"
          alt={formatMessage(messages.dropdownAlt)}
        />
        <Dropdown.Menu>
          <Dropdown.Item
            disabled={isMasquerading}
            onClick={unenrollModal.show}
            data-testid="unenrollModalToggle"
          >
            {formatMessage(messages.unenroll)}
          </Dropdown.Item>
          <Dropdown.Item
            disabled={isMasquerading}
            onClick={emailSettingsModal.show}
            data-testid="emailSettingsModalToggle"
          >
            {formatMessage(messages.emailSettings)}
          </Dropdown.Item>
          {/* Disabled pending PM decision on missing quote param in updated FB api.
            {facebook.isEnabled && (
              <Dropdown.Item>
                <ReactShare.FacebookShareButton
                  url={facebook.shareUrl}
                  quote={formatMessage(messages.shareQuote, {
                    courseName,
                    socialBrand: facebook.socialBrand,
                  })}
                >
                  {formatMessage(messages.shareToFacebook)}
                </ReactShare.FacebookShareButton>
              </Dropdown.Item>
            )}
          */}
          {twitter.isEnabled && (
            <Dropdown.Item>
              <ReactShare.TwitterShareButton
                url={twitter.shareUrl}
                title={formatMessage(messages.shareQuote, {
                  courseName,
                  socialBrand: twitter.socialBrand,
                })}
              >
                {formatMessage(messages.shareToTwitter)}
              </ReactShare.TwitterShareButton>
            </Dropdown.Item>
          )}
        </Dropdown.Menu>
      </Dropdown>
      <UnenrollConfirmModal
        show={unenrollModal.isVisible}
        closeModal={unenrollModal.hide}
        cardId={cardId}
      />
      <EmailSettingsModal
        show={emailSettingsModal.isVisible}
        closeModal={emailSettingsModal.hide}
        cardId={cardId}
      />
    </>
  );
};
CourseCardMenu.propTypes = {
  cardId: PropTypes.string.isRequired,
};

export default CourseCardMenu;
