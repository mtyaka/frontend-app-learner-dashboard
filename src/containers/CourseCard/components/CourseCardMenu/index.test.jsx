import { shallow } from 'enzyme';

import { hooks as appHooks } from 'data/redux';
import { useEmailSettings, useUnenrollData } from './hooks';
import CourseCardMenu from '.';

jest.mock('react-share', () => ({
  FacebookShareButton: () => 'FacebookShareButton',
  TwitterShareButton: () => 'TwitterShareButton',
}));
jest.mock('data/redux', () => ({
  hooks: {
    useCardCourseData: jest.fn(),
    useCardSocialSettingsData: jest.fn(),
    useMasqueradeData: jest.fn(),
  },
}));
jest.mock('./hooks', () => ({
  useEmailSettings: jest.fn(),
  useUnenrollData: jest.fn(),
}));

const props = {
  cardId: 'test-card-id',
};
const defaultEmailSettingsModal = {
  isVisible: false,
  show: jest.fn().mockName('emailSettingShow'),
  hide: jest.fn().mockName('emailSettingHide'),
};
const defaultUnenrollModal = {
  isVisible: false,
  show: jest.fn().mockName('unenrollShow'),
  hide: jest.fn().mockName('unenrollHide'),
};
const defaultSocialShare = {
  facebook: {
    isEnabled: true,
    shareUrl: 'facebook-share-url',
    socialBrand: 'facebook-social-brand',
  },
  twitter: {
    isEnabled: true,
    shareUrl: 'facebook-share-url',
    socialBrand: 'facebook-social-brand',
  },
};
const courseName = 'test-course-name';
let wrapper;

describe('CourseCardMenu', () => {
  beforeEach(() => {
    useEmailSettings.mockReturnValue(defaultEmailSettingsModal);
    useUnenrollData.mockReturnValue(defaultUnenrollModal);
    appHooks.useCardSocialSettingsData.mockReturnValue(defaultSocialShare);
    appHooks.useCardCourseData.mockReturnValue({ courseName });
    appHooks.useMasqueradeData.mockReturnValue({ isMasquerading: false });
  });
  test('snapshot', () => {
    wrapper = shallow(<CourseCardMenu {...props} />);
    expect(wrapper).toMatchSnapshot();
    // expect(wrapper.find('FacebookShareButton').length).toEqual(1);
    expect(wrapper.find('TwitterShareButton').length).toEqual(1);
    expect(wrapper.find({
      'data-testid': 'unenrollModalToggle',
    }).props().disabled).toEqual(false);
    expect(wrapper.find({
      'data-testid': 'emailSettingsModalToggle',
    }).props().disabled).toEqual(false);
  });
  test('snapshot: masquerading', () => {
    appHooks.useMasqueradeData.mockReturnValue({ isMasquerading: true });
    wrapper = shallow(<CourseCardMenu {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find({
      'data-testid': 'unenrollModalToggle',
    }).props().disabled).toEqual(true);
    expect(wrapper.find({
      'data-testid': 'emailSettingsModalToggle',
    }).props().disabled).toEqual(true);
  });
  /*
  test('facebook share disabled', () => {
    appHooks.useCardSocialSettingsData.mockReturnValueOnce({
      ...defaultSocialShare,
      facebook: { ...defaultSocialShare.facebook, isEnabled: false },
    });
    wrapper = shallow(<CourseCardMenu {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('FacebookShareButton').length).toEqual(0);
  });
  */
  test('twitter share disabled', () => {
    appHooks.useCardSocialSettingsData.mockReturnValueOnce({
      ...defaultSocialShare,
      twitter: { ...defaultSocialShare.twitter, isEnabled: false },
    });
    wrapper = shallow(<CourseCardMenu {...props} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('TwitterShareButton').length).toEqual(0);
  });
});
