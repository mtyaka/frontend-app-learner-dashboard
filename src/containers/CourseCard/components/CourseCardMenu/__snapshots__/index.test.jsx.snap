// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseCardMenu facebook share disabled 1`] = `
<Fragment>
  <Dropdown>
    <Dropdown.Toggle
      alt="Course actions dropdown"
      as="IconButton"
      iconAs="Icon"
      id="dropdown-toggle-with-iconbutton"
      src={[MockFunction icons.MoreVert]}
      variant="primary"
    />
    <Dropdown.Menu>
      <Dropdown.Item
        data-testid="unenrollModalToggle"
        disabled={false}
        onClick={[MockFunction unenrollShow]}
      >
        Unenroll
      </Dropdown.Item>
      <Dropdown.Item
        data-testid="emailSettingsModalToggle"
        disabled={false}
        onClick={[MockFunction emailSettingShow]}
      >
        Email settings
      </Dropdown.Item>
      <Dropdown.Item>
        <TwitterShareButton
          title="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
          url="facebook-share-url"
        >
          Share to Twitter
        </TwitterShareButton>
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
  <UnenrollConfirmModal
    cardId="test-card-id"
    closeModal={[MockFunction unenrollHide]}
    show={false}
  />
  <EmailSettingsModal
    cardId="test-card-id"
    closeModal={[MockFunction emailSettingHide]}
    show={false}
  />
</Fragment>
`;

exports[`CourseCardMenu snapshot 1`] = `
<Fragment>
  <Dropdown>
    <Dropdown.Toggle
      alt="Course actions dropdown"
      as="IconButton"
      iconAs="Icon"
      id="dropdown-toggle-with-iconbutton"
      src={[MockFunction icons.MoreVert]}
      variant="primary"
    />
    <Dropdown.Menu>
      <Dropdown.Item
        data-testid="unenrollModalToggle"
        disabled={false}
        onClick={[MockFunction unenrollShow]}
      >
        Unenroll
      </Dropdown.Item>
      <Dropdown.Item
        data-testid="emailSettingsModalToggle"
        disabled={false}
        onClick={[MockFunction emailSettingShow]}
      >
        Email settings
      </Dropdown.Item>
      <Dropdown.Item>
        <FacebookShareButton
          quote="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
          url="facebook-share-url"
        >
          Share to Facebook
        </FacebookShareButton>
      </Dropdown.Item>
      <Dropdown.Item>
        <TwitterShareButton
          title="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
          url="facebook-share-url"
        >
          Share to Twitter
        </TwitterShareButton>
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
  <UnenrollConfirmModal
    cardId="test-card-id"
    closeModal={[MockFunction unenrollHide]}
    show={false}
  />
  <EmailSettingsModal
    cardId="test-card-id"
    closeModal={[MockFunction emailSettingHide]}
    show={false}
  />
</Fragment>
`;

exports[`CourseCardMenu snapshot: masquerading 1`] = `
<Fragment>
  <Dropdown>
    <Dropdown.Toggle
      alt="Course actions dropdown"
      as="IconButton"
      iconAs="Icon"
      id="dropdown-toggle-with-iconbutton"
      src={[MockFunction icons.MoreVert]}
      variant="primary"
    />
    <Dropdown.Menu>
      <Dropdown.Item
        data-testid="unenrollModalToggle"
        disabled={true}
        onClick={[MockFunction unenrollShow]}
      >
        Unenroll
      </Dropdown.Item>
      <Dropdown.Item
        data-testid="emailSettingsModalToggle"
        disabled={true}
        onClick={[MockFunction emailSettingShow]}
      >
        Email settings
      </Dropdown.Item>
      <Dropdown.Item>
        <FacebookShareButton
          quote="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
          url="facebook-share-url"
        >
          Share to Facebook
        </FacebookShareButton>
      </Dropdown.Item>
      <Dropdown.Item>
        <TwitterShareButton
          title="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
          url="facebook-share-url"
        >
          Share to Twitter
        </TwitterShareButton>
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
  <UnenrollConfirmModal
    cardId="test-card-id"
    closeModal={[MockFunction unenrollHide]}
    show={false}
  />
  <EmailSettingsModal
    cardId="test-card-id"
    closeModal={[MockFunction emailSettingHide]}
    show={false}
  />
</Fragment>
`;

exports[`CourseCardMenu twitter share disabled 1`] = `
<Fragment>
  <Dropdown>
    <Dropdown.Toggle
      alt="Course actions dropdown"
      as="IconButton"
      iconAs="Icon"
      id="dropdown-toggle-with-iconbutton"
      src={[MockFunction icons.MoreVert]}
      variant="primary"
    />
    <Dropdown.Menu>
      <Dropdown.Item
        data-testid="unenrollModalToggle"
        disabled={false}
        onClick={[MockFunction unenrollShow]}
      >
        Unenroll
      </Dropdown.Item>
      <Dropdown.Item
        data-testid="emailSettingsModalToggle"
        disabled={false}
        onClick={[MockFunction emailSettingShow]}
      >
        Email settings
      </Dropdown.Item>
      <Dropdown.Item>
        <FacebookShareButton
          quote="I'm taking test-course-name online with facebook-social-brand.  Check it out!"
          url="facebook-share-url"
        >
          Share to Facebook
        </FacebookShareButton>
      </Dropdown.Item>
    </Dropdown.Menu>
  </Dropdown>
  <UnenrollConfirmModal
    cardId="test-card-id"
    closeModal={[MockFunction unenrollHide]}
    show={false}
  />
  <EmailSettingsModal
    cardId="test-card-id"
    closeModal={[MockFunction emailSettingHide]}
    show={false}
  />
</Fragment>
`;
